planner/cascades/transformation_rules.go

```diff
@@ -634,3 +635,51 @@ func (r *EliminateProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.
 	}
 	return finalGroupExprs, true, false, nil
 }
+
+// MergeAdjacentProjection merge the adjacent projection.
+type MergeAdjacentProjection struct {
+	baseRule
+}
+
+// NewRuleMergeAdjacentProjection creates a new Transformation MergeAdjacentProjection.
+// The pattern of this rule is `Projection -> Projection`.
+func NewRuleMergeAdjacentProjection() Transformation {
+	rule := &MergeAdjacentProjection{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandProjection,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandProjection, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// OnTransform implements Transformation interface.
+// It will transform `proj->proj->x` to `proj->x`
+// or just keep the adjacent projections unchanged.
+func (r *MergeAdjacentProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	proj := old.GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	childGroup := old.Children[0].Group
+	child := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	if plannercore.ExprsHasSideEffects(child.Exprs) {
+		return nil, false, false, nil
+	}
+
+	replace := make(map[string]*expression.Column)
+	for i, col := range childGroup.Prop.Schema.Columns {
+		if colOrigin, ok := child.Exprs[i].(*expression.Column); ok {
+			replace[string(col.HashCode(nil))] = colOrigin
+		}
+	}
+
+	newProj := plannercore.LogicalProjection{Exprs: make([]expression.Expression, len(proj.Exprs))}.Init(proj.SCtx(), proj.SelectBlockOffset())
+	copy(newProj.Exprs, proj.Exprs)
+	newProj.SetSchema(old.GetExpr().Group.Prop.Schema)
+	for i, expr := range newProj.Exprs {
+		plannercore.ResolveExprAndReplace(expr, replace)
+		newProj.Exprs[i] = plannercore.ReplaceColumnOfExpr(newProj.Exprs[i], child, childGroup.Prop.Schema)
+	}
```

Comment:
`copy(newExprs, oldExprs)` is not safe enough, `ReplaceColumnOfExpr` will change the `args` of origin `ScalarFunction`. Try use a deep clone for each expression:
```
for i := range newProj.Exprs {
    newProj.Exprs[i] = newProj.Exprs[i].Clone()    
}
```